contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,212 @@
+
+Only cluster admins will be able to define a ResourceClass by default. This can
+then be tweaked through RBAC.
+
+A ResourceClass spec will have the following spec:
+```yaml
+kind: ResourceClass
+apiVersion: compute.k8s.io/v1alpha1
+metadata:
+  name: nvidia-tesla-3.7
+spec:
+    resourceSelector:  // required are ORed
+        - required:  // all properties under required are ANDed
+          - key: kind
+            operator: In
+            values: ["nvidia-gpu"]
+          - key: memory
+            operator: Gt
+            values: ["8 Gi"]
+        - required:
+          - key: kind
+            operator: In
+            values: ["nvidia-gpu"]
+          - key: compute-capabilities
+            operator: Gt
+            values: ["3.7"]
+
+```
+
+The user will then be able to specify in his pod Spec requirements a
+Resource class through either Requests or Limits
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: gpu-pod
+spec:
+  containers:
+    name: gpu-container
+    resources:
+      limits:
+        cpu: 1.2
+        memory: 10Gi
+        nvidia-tesla-3.7: 4
+      requests:
+        cpu: 0.3
+```
+
+
+### Device Plugin
+
+Kubelet will create a gRPC connection (as a client) to the Device Plugins whose endpoint
+were specified on the CLI.
+
+The kubelet will call the `Discover` function once and then listen to a stream returned
+by the `Monitor` function.
+
+When receiving a pod which requests GPUs (through the ResourceClass api) kubelet will
+be in charge of deciding which device to assign and advertising the corresponding changes
+to the node's allocatable pool and pod assigned devices.
+The scheduler will still be in charge of filtering the nodes which cannot satisfy the
+requests.
+
+The typical process will follow the following pattern:
+1. A user submits a pod spec requesting X devices with Y properties
+2. The scheduler filters the nodes which do not match the requests
+3. The pod lands on the node
+4. Kubelet decides which device should be assigned to the pod
+5. Kubelet removes the devices from the `Allocatable` pool
+6. Kubelet updates the pod's status with the allocated devices
+7. Kubelet calls `Allocate` on the matching Device Plugins
+8. The user deletes the pod or the pod terminates
+9. Kubelet calls `Deallocate` on the matching Device Plugins
+10. Kubelet puts the devices back in the `Allocatable` pool
+
+![Process](./device-plugin.png)
+
+The kubelet will also be able to call `Allocate` and `Deallocate` on any devices returned
+by the `Discover` function.
+
+Calling `Allocate` will return a CRI spec which allows the plugin to set cgroups, specify
+environment variables, ...
+
+Calling `Allocate` on a device already Allocated and not Deallocated as well as
+calling `Deallocate` on a device not Allocated or already Deallocated should return an error.
+
+When calling Allocate or Deallocate, only the Name field needs to be set.
+
+```go
+service DeviceManager {
+	rpc Discover() returns (stream ResourceType);
```

Comment:
There needs to be separate API types for this interface. The public k8s APIs should not be re-used here ideally for having clean, modular, separate abstractions